### PR TITLE
[Domain] Build AccountBalanceUseCase

### LINKED ISSUE #3143: [Domain] Build the `AccountBalanceUseCase`
### Please confirm the following

- [X] I checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.

### What would you like to improve?

- Build the `AccountBalanceUseCase`.
- Use the `AccountStatsUseCase` and the `TransactionRepository`

```kotlin
class AccountBalanceUseCase @Inject constructor(...) {
   fun calculate(account: AccountId): Map<AssetCode, NonZeroDouble>
}
```

- Build the `NonZeroDouble` following the pattern to the `PositiveDouble`
- We neee NonZeroDouble to handle the case of negative balances (credits to @Vishwa-Raghavendra)

### Because...

- Proper domain logic following the architecture and having unit + property-based tests.

### Description

_No response_

### Success Criteria

- Follows all dev guidelines (this is a critical code so expect Requests for changes on your PR)
- Property-based tests
- Unit tests

### PR DESCRIPTION
**Pull request by rodrigoliveirac**
## Pull request (PR) checklist

Please check if your pull request fulfills the following requirements:

- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read and understand the [Developer Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Guidelines.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] My PR includes only the necessary changes to fix the issue (i.e., no unnecessary files or lines of code are changed).
- [x] 🎬 I've attached a **screen recording** of using the new code to the next paragraph (if applicable).

## What's changed?

Describe with a few bullets **what's new:**
- I've implemented the methods of AccountBalanceUseCase

## Does this PR close any GitHub issues? (do not delete)

- Closes #3143

## Troubleshooting GitHub Actions (CI) failures ❌
Pull request checks failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 1]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+        val amountIn = income.values.map {
+            it.key to transfersIn.values[it.key]?.value?.plus(it.value.value)
+        }.toMap()
+
+        val amountOut = expense.values.map {
+            it.key to transfersOut.values[it.key]?.value?.plus(it.value.value)
+        }.toMap()
+
+        amountIn.map {
+            balance[it.key] = NonZeroDouble.unsafe(it.value?.minus(amountOut[it.key]!!)!!)
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:62]: This looks unsafe and will likely produce runtime error => crashes

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 2]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
     suspend fun calculate(
         account: AccountId,
     ): Map<AssetCode, NonZeroDouble> {
-        TODO("Not implemented")
+
+        val transactions = transactionRepository.findAll()
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:43]: inline the variable and make the below line call on 2 lines, will be easier to read IMO

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 3]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+
+        val amountIn = income.values.map {
+            it.key to transfersIn.values[it.key]?.value?.plus(it.value.value)
+        }.toMap()
+
+        val amountOut = expense.values.map {
+            it.key to transfersOut.values[it.key]?.value?.plus(it.value.value)
+        }.toMap()
+
+        amountIn.map {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:61]: That's not mapping because it seems that you don't use the result. Replace with `.forEach`

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 1 above)
Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:62]: This will crash for the case of 0 when transfersIn = transfersOut

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 4]
```diff
@@ -23,7 +23,13 @@ class AccountBalanceUseCase @Inject constructor(
         account: AccountId,
         outCurrency: AssetCode
     ): ExchangedAccountBalance {
-        TODO("Not implemented")
+        val balance = calculate(account)
+        val exchangeResult = exchangeUseCase.convert(values = balance, to = outCurrency)
+        return if(balance.keys.size == exchangeResult.exchangeErrors.size || balance.values.sumOf { it.value } == 0.0) {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: I don't understand this `if`. What's its purpose?

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 5]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+        val income = accountStats.income
+        val transfersIn = accountStats.transfersIn
+        val expense = accountStats.expense
+        val transfersOut = accountStats.transfersOut
+
+        val amountIn = income.values.map {
+            it.key to transfersIn.values[it.key]?.value?.plus(it.value.value)
+        }.toMap()
+
+        val amountOut = expense.values.map {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:57]: This logic is wrong: it won't include transfersOut in different currency than the expenses.

E.g.:
- 10 BGN expense
- 20 USD transfer out

=> amountOut will be just 10 BGN

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
Found some issues with the logic + crashes. The code can be also be simplified a bit

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 6]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+
+        val balance = hashMapOf<AssetCode, NonZeroDouble>()
+
+        val income = accountStats.income
+        val transfersIn = accountStats.transfersIn
+        val expense = accountStats.expense
+        val transfersOut = accountStats.transfersOut
+
+        val amountIn = income.values.map {
+            it.key to transfersIn.values[it.key]?.value?.plus(it.value.value)
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:54]: Use `(asset, amount) ->` instead of `it`. The current version is hard to read and understand

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 4 above)
Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: I was unsure about that. The docs said to return no balance if the balance is zero or if the exchange to outCurrency failed for all assets. So, if the first case occurs, the sum of the values should be 0.0. If the second occurs, the exchangeErrors.size should be equal to the size of all asset codes, right?

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 5 above)
Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:57]: I see. I'll fix the logic and improve the code quality. Thanks!

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 4 above)
Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: > If the second occurs, the exchangeErrors.size should be equal to the size of all asset codes, right?

E.g. 2 assets and 2 errors doesn't mean that we should return empty. IMO, this if is unnecessary. The user of the API should check for themselves if there are errors and how to handle them

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 7]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
     }
 }
 
+
 data class ExchangedAccountBalance(
-    val balance: Option<Value>,
-    val exchangeErrors: Set<AssetCode>
-)
\ No newline at end of file
+    val balance: Option<Value>?,
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:71]: This shouldn't be null. Use `Option.NONE`

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 8]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+    val exchangeErrors: Set<AssetCode>?
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:72]: This shouldn't be null. If there are no errors => just empty set

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 4 above)
Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: The check should be:

if balances.isEmpty => balance = Option.None

Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: errors stay as they are

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 9]
```diff
@@ -33,11 +39,42 @@ class AccountBalanceUseCase @Inject constructor(
...
+) {
+    companion object {
+        val None = ExchangedAccountBalance(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:75]: balance = Option.None, exchangeErrors = emptySets()

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt (hunk 4 above)
Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:28]: Got it. Thanks!

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
+
+    private val balance = mutableMapOf<AssetCode, NonZeroDouble>()
+
+    fun processIncomes(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:11]: processAssetsIn + add some KDoc explaining what it does

Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:11]: actually nvm, it's a bad name suggestion

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        combineAndGetAmount(incomes, transferIn).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + value.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processOutcomes(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:24]: Is there a better name than "outcomes"? We should check the banking term for those. Like debits (money out) and credits (money in)

Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt]: I found this link https://www.investopedia.com/terms/w/withdrawal.asp to use for outcomes and "deposit" for incomes in https://www.bankofamerica.com/deposits/manage/glossary/

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt (hunk 10 above)
Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt]: Ok

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    private fun combine(
+        amount: MutableMap<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>
+    ) {
+        secondMap.forEach { (asset, value) ->
+            amount.merge(asset, value) { firstMap, secondMap ->
+                PositiveDouble
+                    .from(firstMap.value + secondMap.value)
+                    .fold(
+                        ifLeft = { PositiveDouble.unsafe(0.0) },
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:55]: `PositiveDouble.unsafe(0.0)` is 100% crash. The idea of PositiveDouble and other exact types is to guarantee some property (check the class impl). You should **never** use `unsafe` unless you can mathematically proof that the given property persists. Check [this docs](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/guidelines/Data-Modeling.md#data-modeling) - feel free to suggest improvements on top of it

You should use `PositiveDouble.from()`. Since `0.0` is not positive => that's not a positive double. It's a `NonNegativeDouble`

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        combine(amount, secondMap)
+        return amount
+    }
+
+    private fun combine(
+        amount: MutableMap<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>
+    ) {
+        secondMap.forEach { (asset, value) ->
+            amount.merge(asset, value) { firstMap, secondMap ->
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:51]: This namings are confusing. Let's call the two maps `a` and `b` or something consistent. Now `firstMap` and `amount` are clashing and I can't easily understand the code

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
+
+    private val balance = mutableMapOf<AssetCode, NonZeroDouble>()
+
+    fun processDeposit(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:11]: This namings are good 💯 nice!

Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:11]: Let's rename it to `processDeposits` and `processWithdrawals` (plural because we're processing more than once thing)

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        getAmount(incomes, transferIn).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + value.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processWithdrawal(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:24]: `processWithdrawals` (plural) sound better

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        getAmount(expenses, transferOut).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + (-value.value))
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    private fun getAmount(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:37]: That's `combine` tbh. Just copy-paste the contents of `combine` here and we can end-up with just one function. Also `a` and `b` IMO are better names for the maps. From math POV:
- combine: a + b -> c

Combines two arbitrary maps into a third one

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processWithdrawal(
+        expenses: Map<AssetCode, PositiveDouble>,
+        transferOut: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(expenses, transferOut).forEach { (asset, value) ->
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:28]: `-value.value` reads badly, rename the 2nd param to `(asset, amount) ->` this way it'll read better:
`amount.value`

**Comment by ILIYANGERMANOV** [comment]
Hey @rodrigoliveirac sorry for all the comments and nitpicking! It's just a core API that we'll use a lot and I want us to make it as simple and elegant as possible 🙌  Btw, great progress - I like the new builder + naming!

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt (hunk 16 above)
Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:37]: Fair enough!

**Comment by rodrigoliveirac** [comment]
> Hey @rodrigoliveirac sorry for all the comments and nitpicking! It's just a core API that we'll use a lot and I want us to make it as simple and elegant as possible 🙌 Btw, great progress - I like the new builder + naming!

No need to apologize! I definitely understand the importance of that. I'm enjoying the opportunity to grow with this. Thank you for all!

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + (-value.value))
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:30]: Can be simplified to:
```
(balance[asset]?.value ?: 0.0) - amount.value
```

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 19]
```diff
@@ -0,0 +1,63 @@
+package com.ivy.domain.usecase
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:7]: You can start with unit tests for this class. See the ones that I did for the other StatSummaryBuilder

Consider adding:
- unit tests
- property based tests verifying Math properties that you can identify (e.g. the order of transactions doesn't matter for the balance)

Comment (rodrigoliveirac) [reply shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:7]: Ok. I'll read about `PositiveDouble` to solve that Left case and then move on to the tests

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * Updates `balance` with deposits from income and transfers in.
+     *
+     * @param incomes Deposit amounts from income sources.
+     * @param transferIn Deposit amounts from transfer in sources.
+     */
+    fun processDeposits(
+        incomes: Map<AssetCode, PositiveDouble>,
+        transferIn: Map<AssetCode, PositiveDouble>,
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:19]: Rename to plural `transfersIn` because there can be many transfers

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * Updates `balance` with withdrawals from expenses and transfers out.
+     *
+     * @param expenses Deposit amounts from expense sources.
+     * @param transferOut Deposit amounts from transfer out sources.
+     */
+    fun processWithdrawals(
+        expenses: Map<AssetCode, PositiveDouble>,
+        transferOut: Map<AssetCode, PositiveDouble>,
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:38]: Rename to plural

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+                .from((balance[asset]?.value ?: 0.0) - amount.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    /**
+     * Combines two maps by summing their values. If the sum results in an error,
+     * the value from the `c` map is used.
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:51]: Users of the API don't know about `c` - isn't the value of `a` used?

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 23]
```diff
@@ -33,11 +44,35 @@ class AccountBalanceUseCase @Inject constructor(
...
         account: AccountId,
     ): Map<AssetCode, NonZeroDouble> {
-        TODO("Not implemented")
+        val accountStats = accountStatsUseCase.calculate(
+            account = account,
+            transactions = transactionRepository.findAll()
+        )
+
+        val balance = BalanceBuilder()
+
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:53]: unnecessary blank line - blank lines can help readability when separating logically different operations

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 24]
```diff
@@ -33,11 +44,35 @@ class AccountBalanceUseCase @Inject constructor(
...
+        balance.processDeposits(
+            incomes = accountStats.income.values,
+            transferIn = accountStats.transfersIn.values
+        )
+
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:58]: unnecessary blank line

File: shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt [hunk 25]
```diff
@@ -33,11 +44,35 @@ class AccountBalanceUseCase @Inject constructor(
     suspend fun calculate(
         account: AccountId,
     ): Map<AssetCode, NonZeroDouble> {
-        TODO("Not implemented")
+        val accountStats = accountStatsUseCase.calculate(
+            account = account,
+            transactions = transactionRepository.findAll()
+        )
+
+        val balance = BalanceBuilder()
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/account/AccountBalanceUseCase.kt:52]: To make it more Kotlin idiomatic we can:
```kotlin
return BalanceBuilder().run {
   processDeposits(...)
   processWithdtwqlas(...)
   build()
}
```

The code would be shorter and read nicer IMO

**Comment by ILIYANGERMANOV** [comment]
Nice work @rodrigoliveirac 👏 I really start liking this implementation! What's left is cover all cases with tests and fix edge-case. Trust me, property-based tests will uncover ones

**Comment by rodrigoliveirac** [comment]
> Nice work @rodrigoliveirac 👏 I really start liking this implementation! What's left is cover all cases with tests and fix edge-case. Trust me, property-based tests will uncover ones

Sure! I'll do. Thank you! 👊

**Comment by rodrigoliveirac** [comment]
@ILIYANGERMANOV Is this test structure okay?

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+        val balance = balanceBuilder.build()
+
+        // then
+        balance shouldBe testCase.expected
+    }
+
+    @Test
+    fun `process one deposit come from incomes`() {
+        // given
+        val testCase = ValuesTestCase.OneDepositFromIncome
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:200]: Use `@TestParameter` and `@RunWith(TestParameterInjector::class)` search it in my other tests

Comment (ILIYANGERMANOV) [reply shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:200]: There should be one test with multiple test cases

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    @Test
+    fun `process two deposits in different currencies come from incomes and transfersIn`() {
+        // given
+        val testCase = ValuesTestCase.TwoInDiffCurrencyDepositsFromIncomeAndTransfersIn
+        val balanceBuilder = BalanceBuilder()
+
+        // when
+        val stats = testCase.values
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:261]: That's part of //given

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:262]: unnecessary empty line

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt [hunk 29]
```diff
@@ -0,0 +1,300 @@
...
+import com.ivy.data.model.primitive.NonNegativeInt
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+import com.ivy.domain.model.StatSummary
+import com.ivy.domain.usecase.BalanceBuilder
+import com.ivy.domain.usecase.account.AccountStats
+import io.kotest.matchers.shouldBe
+import org.junit.Test
+
+class BalanceBuilderTest {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:16]: You can also building another test class BalanceBuilderPropertyTest

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt [hunk 30]
```diff
@@ -0,0 +1,300 @@
...
+    @Test
+    fun `process one deposit come from incomes`() {
+        // given
+        val testCase = ValuesTestCase.OneDepositFromIncome
+        val balanceBuilder = BalanceBuilder()
+
+        // when
+        val stats = testCase.values
+
+        balanceBuilder.processDeposits(
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderTest.kt:206]: All test cases should both process deposits and withdrawals
